ddr/lib/ddr-ir/Macro.cpp

```diff
@@ -25,57 +25,951 @@
+	}
+	return false;
+}
+
+/**
+ * Evaluates the addition expression (a + b), and the subtraction expression
+ * (a - b), which have the same operator precedence, by recursively calling
+ * the functions for evaluating expressions with higher precedence.
+ *
+ * @param[out] ret: the integer value the logical and expression evaluates to
+ *
+ * @return: whether the at the current index the expression can be evaluated
+ */
+bool
+MacroParser::addSub(int64_t *ret)
+{
+	int64_t term1 = 0;
+	int64_t term2 = 0;
+
+	if (multDivRem(&term1)) {
+		while (_scanner._cursor < _scanner._end) {
+			if (_scanner.atSymbol(PLUS)) {
+				if (multDivRem(&term2)) {
+					term1 = term1 + term2;
+				} else {
+					return false;
+				}
+			} else if (_scanner.atSymbol(MINUS)) {
+				if (multDivRem(&term2)) {
+					term1 = term1 - term2;
+				} else {
+					return false;
+				}
+			} else {
+				break;
+			}
+		}
+		*ret = term1;
+		return true;
+	}
+	return false;
+}
+
+/**
+ * Evaluates the multiplication (a * b), division (a / b), and remainder
+ * (a % b) expressions, which have the same operator precedence, by recursively
+ * calling the functions for evaluating expressions with higher precedence.
+ * Since this is the highest precedence binary expression, it will attempt to
+ * evaluate its operands as either unary expressions, parenthesized
+ * expressions, or numbers.
+ *
+ * @param[out] ret: the integer value the logical and expression evaluates to
+ *
+ * @return: whether the at the current index the expression can be evaluated
+ */
+bool
+MacroParser::multDivRem(int64_t *ret)
+{
+	int64_t term1 = 0;
+	int64_t term2 = 0;
+
+	if (unaryExpression(&term1) || parentheses(&term1)
+			|| _scanner.readNumber(&term1)) {
+		while (_scanner._cursor < _scanner._end) {
+			if (_scanner.atSymbol(MULT)) {
+				if (unaryExpression(&term2) || parentheses(&term2)
+						|| _scanner.readNumber(&term2)) {
+					term1 = term1 * term2;
+				} else {
+					return false;
+				}
+			} else if (_scanner.atSymbol(DIV)) {
+				if (unaryExpression(&term2) || parentheses(&term2)
+						|| _scanner.readNumber(&term2)) {
+					term1 = term1 / term2;
+				} else {
+					return false;
+				}
+			} else if (_scanner.atSymbol(REM)) {
+				if (unaryExpression(&term2) || parentheses(&term2)
+						|| _scanner.readNumber(&term2)) {
+					term1 = term1 % term2;
+				} else {
+					return false;
+				}
+			} else {
+				break;
+			}
+		}
+		*ret = term1;
+		return true;
+	}
+	return false;
+}
+
+/**
+ * helper function for unaryExpression() function, parses the types in and
+ * applies a unary typecast operator
+ *
+ * @param[in] typeExpr: string representing the types inside a typecast operator
+ * @param[in] operand: value of the operand the typecast is being applied to
+ * @param[out] ret: value after the typecast has been applied
+ *
+ * @return: whether the typecast is valid and has been successfully applied
+ */
+static bool
+typeCast(const string typeExpr, int64_t operand, int64_t *ret)
```

Comment:
I think we should first just ignore casts (besides, this doesn't handle things like `uintptr_t` or `UDATA`).

Removing the friendship between the scanner and parser classes will require some rework. I expect the scanner will need to recognize a sequence of words separated by whitespace for use by the parser to handle casts.